Bug: missing nil check when node is deleted, but connections not yet cleaned

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102dbfb98]

goroutine 9793 [running]:
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*vertex).isUnderMaintenance(0x14000189680?)
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/vertex.go:28 +0x18
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).knnSearchByVector(0x14000189680, {0x14037dab800, 0x180, 0x180}, 0x64, 0x1611d378?, 0x1611d37800000000?)
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/search.go:412 +0x4b8
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).SearchByVector(0x14000189680, {0x1411c508800, 0x180, 0x1da}, 0x0?, 0x0)
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/search.go:74 +0x128
github.com/semi-technologies/weaviate/adapters/repos/db.(*Shard).objectVectorSearch(0x1400058ac40, {0x1039fec00, 0x1411c3cc480}, {0x1411c508800, 0x180, 0x1da}, 0x26e294c?, 0x64, 0x0, {0x0, ...}, ...)
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/adapters/repos/db/shard_read.go:222 +0x160
github.com/semi-technologies/weaviate/adapters/repos/db.(*Index).objectVectorSearch.func1()
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/adapters/repos/db/index.go:799 +0x180
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/vendor/golang.org/x/sync/errgroup/errgroup.go:75 +0x5c
created by golang.org/x/sync/errgroup.(*Group).Go
        /Users/etiennedilocker/code/github.com/semi-technologies/weaviate/vendor/golang.org/x/sync/errgroup/errgroup.go:72 +0xa4
exit status 2
```

This seems to be when a node has been "partially cleaned up" after being deleted.

--- a/adapters/repos/db/vector/hnsw/search.go
+++ b/adapters/repos/db/vector/hnsw/search.go
@@ -407,9 +407,12 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
 		// that particular level, so instead we're keeping whatever entrypoint we
 		// had before (i.e. either from a previous level or even the main
 		// entrypoint)
+		//
+		// If we do, however, have results, any candidate that's not nil (not
+		// deleted), and not under maintenance is a viable candidate
 		for res.Len() > 0 {
 			cand := res.Pop()
-			if !h.nodeByID(cand.ID).isUnderMaintenance() {
+			if n := h.nodeByID(cand.ID); n != nil && !n.isUnderMaintenance() {
 				entryPointID = cand.ID
 				entryPointDistance = cand.Dist
 				break

--- a/adapters/repos/db/vector/hnsw/search.go
+++ b/adapters/repos/db/vector/hnsw/search.go
@@ -412,7 +412,20 @@ func (h *hnsw) knnSearchByVector(searchVec []float32, k int,
 		// deleted), and not under maintenance is a viable candidate
 		for res.Len() > 0 {
 			cand := res.Pop()
-			if n := h.nodeByID(cand.ID); n != nil && !n.isUnderMaintenance() {
+			n := h.nodeByID(cand.ID)
+			if n == nil {
+				// we have found a node in results that is nil. This means it was
+				// deleted, but not cleaned up properly. Make sure to add a tombstone to
+				// this node, so it can be cleaned up in the next cycle.
+				if err := h.addTombstone(cand.ID); err != nil {
+					return nil, nil, err
+				}
+
+				// skip the nil node, as it does not make a valid entrypoint
+				continue
+			}
+
+			if !n.isUnderMaintenance() {
 				entryPointID = cand.ID
 				entryPointDistance = cand.Dist
 				break

--- a/adapters/repos/db/vector/hnsw/search_test.go
+++ b/adapters/repos/db/vector/hnsw/search_test.go
@@ -1,0 +1,77 @@
+package hnsw
+
+import (
+	"context"
+	"testing"
+
+	"github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/distancer"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// prevents a regression of
+// https://github.com/semi-technologies/weaviate/issues/2155
+func TestNilCheckOnPartiallyCleanedNode(t *testing.T) {
+	vectors := [][]float32{
+		{100, 100}, // first to import makes this the EP, it is far from any query which means it will be replaced.
+		{2, 2},     // a good potential entrypoint, but we will corrupt it later on
+		{1, 1},     // the perfect search result
+	}
+
+	var vectorIndex *hnsw
+
+	t.Run("import", func(*testing.T) {
+		index, err := New(Config{
+			RootPath:              "doesnt-matter-as-committlogger-is-mocked-out",
+			ID:                    "bug-2155",
+			MakeCommitLoggerThunk: MakeNoopCommitLogger,
+			DistanceProvider:      distancer.NewL2SquaredProvider(),
+			VectorForIDThunk: func(ctx context.Context, id uint64) ([]float32, error) {
+				return vectors[int(id)], nil
+			},
+		}, UserConfig{
+			MaxConnections: 30,
+			EFConstruction: 128,
+
+			// The actual size does not matter for this test, but if it defaults to
+			// zero it will constantly think it's full and needs to be deleted - even
+			// after just being deleted, so make sure to use a positive number here.
+			VectorCacheMaxObjects: 100000,
+		})
+		require.Nil(t, err)
+		vectorIndex = index
+	})
+
+	t.Run("manually add the nodes", func(t *testing.T) {
+		vectorIndex.entryPointID = 0
+		vectorIndex.currentMaximumLayer = 1
+		vectorIndex.nodes = []*vertex{
+			{
+				// must be on a non-zero layer for this bug to occur
+				level: 1,
+				connections: [][]uint64{
+					{1, 2},
+					{1},
+				},
+			},
+			nil, // corrupt node
+			{
+				level: 0,
+				connections: [][]uint64{
+					{0, 1, 2},
+				},
+			},
+		}
+	})
+
+	t.Run("run a search that would typically find the new ep", func(t *testing.T) {
+		res, _, err := vectorIndex.SearchByVector([]float32{1.7, 1.7}, 20, nil)
+		require.Nil(t, err)
+		assert.Equal(t, []uint64{2, 0}, res, "right results are found")
+	})
+
+	t.Run("the corrupt node is now marked deleted", func(t *testing.T) {
+		_, ok := vectorIndex.tombstones[1]
+		assert.True(t, ok)
+	})
+}

